components/notifier/src/strategies/notification_strategy.py

```diff
@@ -0,0 +1,67 @@
+        """Return the reports that have a webhook and metrics that require notifying."""
+        notifications = []
+        for report in json["reports"]:
+            notable_metrics = []
+            for subject in report["subjects"].values():
+                for metric_uuid, metric in subject["metrics"].items():
+                    notable = self.check_if_metric_is_notable(metric, metric_uuid, most_recent_measurement_seen)
+                    if notable:
+                        notable_metrics.append(MetricNotificationData(metric, self.data_model, notable))
+            if notable_metrics:
+                for destination_uuid, destination in report.get("notification_destinations", {}).items():
+                    notifications.append(Notification(report, notable_metrics, destination_uuid, destination))
+        return notifications
+
+    def check_if_metric_is_notable(self, metric, metric_uuid, most_recent_measurement_seen) -> Optional[str]:
+        """Determine whether a notification should be generated for the given metric."""
+        if (
+            len(metric["recent_measurements"]) > 0
+            and "status_start" in metric["recent_measurements"][-1][metric["scale"]]
+            and self.long_unchanged_status(metric, metric_uuid, most_recent_measurement_seen)
+        ):
+            return "status_long_unchanged"
+        if self.status_changed(metric, most_recent_measurement_seen):
+            if self.already_notified.__contains__(metric_uuid):
+                self.already_notified.remove(metric_uuid)
+            return "status_changed"
+        return None
+
+    @staticmethod
+    def status_changed(metric, most_recent_measurement_seen: str) -> bool:
+        """Determine if a metric got a new status after the given timestamp."""
+        recent_measurements = metric.get("recent_measurements") or []
+        if len(recent_measurements) < 2:
+            return False  # If there are fewer than two measurements, the metric couldn't have recently changed status
+        scale = metric["scale"]
+        metric_had_other_status = recent_measurements[-2][scale]["status"] != recent_measurements[-1][scale]["status"]
+        change_was_recent = recent_measurements[-1]["start"] > most_recent_measurement_seen
+        return bool(metric_had_other_status and change_was_recent)
+
+    def long_unchanged_status(self, metric, metric_uuid, most_recent_measurement_seen: str) -> bool:
+        """Determine if a metric has been red for 3 weeks."""
+        status_start = datetime.fromisoformat(metric["status_start"])
+        datetime_most_recent_measurement_seen = datetime.fromisoformat(most_recent_measurement_seen)
+        difference = datetime_most_recent_measurement_seen - status_start
+        if timedelta(days=21).total_seconds() < difference.total_seconds() < timedelta(
+            days=22
+        ).total_seconds() and not self.already_notified.__contains__(metric_uuid):
```

Comment:
Timedelta's can be compared directly. Also, don't use __contains__ directly, use "in":
```suggestion
        if timedelta(days=21) < difference < timedelta(days=22) and metric_uuid not in self.already_notified:
```